<a href="https://colab.research.google.com/github/mynhungg/Datamining/blob/main/PhoBERT_NLP.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [36]:
from google.colab import drive
import os

drive.flush_and_unmount()

In [37]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [38]:
from google.colab import drive
import os

drive.mount('/content/gdrive', force_remount=True)
path = '/content/gdrive/MyDrive/PhoBERT_NLP'
os.chdir(path)
!ls

Mounted at /content/gdrive
label_test.pkl	model_ckpt_full_both  text_test.pkl


In [39]:
os.chdir('/content')

In [40]:
!pip3 install fairseq fastbpe vncorenlp transformers sentencepiece

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [41]:
!wget https://raw.githubusercontent.com/stopwords/vietnamese-stopwords/master/vietnamese-stopwords-dash.txt

--2023-05-25 02:50:26--  https://raw.githubusercontent.com/stopwords/vietnamese-stopwords/master/vietnamese-stopwords-dash.txt
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 20475 (20K) [text/plain]
Saving to: ‘vietnamese-stopwords-dash.txt.1’

vietnamese-stopword 100%[===================>]  20.00K  --.-KB/s    in 0.001s  

2023-05-25 02:50:26 (34.6 MB/s) - ‘vietnamese-stopwords-dash.txt.1’ saved [20475/20475]



In [42]:
!wget https://public.vinai.io/PhoBERT_base_fairseq.tar.gz
!tar -xzvf PhoBERT_base_fairseq.tar.gz
# !wget https://public.vinai.io/PhoBERT_base_transformers.tar.gz
# !tar -xzvf PhoBERT_base_transformers.tar.gz


--2023-05-25 02:50:27--  https://public.vinai.io/PhoBERT_base_fairseq.tar.gz
Resolving public.vinai.io (public.vinai.io)... 52.84.162.15, 52.84.162.17, 52.84.162.36, ...
Connecting to public.vinai.io (public.vinai.io)|52.84.162.15|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1243308020 (1.2G) [application/x-tar]
Saving to: ‘PhoBERT_base_fairseq.tar.gz.1’

PhoBERT_base_fairse 100%[===================>]   1.16G   119MB/s    in 9.2s    

2023-05-25 02:50:36 (129 MB/s) - ‘PhoBERT_base_fairseq.tar.gz.1’ saved [1243308020/1243308020]

PhoBERT_base_fairseq/
PhoBERT_base_fairseq/bpe.codes
PhoBERT_base_fairseq/model.pt
PhoBERT_base_fairseq/dict.txt


In [44]:
!mkdir -p vncorenlp/models/wordsegmenter
!wget https://raw.githubusercontent.com/vncorenlp/VnCoreNLP/master/VnCoreNLP-1.1.1.jar
!wget https://raw.githubusercontent.com/vncorenlp/VnCoreNLP/master/models/wordsegmenter/vi-vocab
!wget https://raw.githubusercontent.com/vncorenlp/VnCoreNLP/master/models/wordsegmenter/wordsegmenter.rdr
!mv VnCoreNLP-1.1.1.jar vncorenlp/ 
!mv vi-vocab vncorenlp/models/wordsegmenter/
!mv wordsegmenter.rdr vncorenlp/models/wordsegmenter/

--2023-05-25 02:51:12--  https://raw.githubusercontent.com/vncorenlp/VnCoreNLP/master/VnCoreNLP-1.1.1.jar
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 27412575 (26M) [application/octet-stream]
Saving to: ‘VnCoreNLP-1.1.1.jar’

VnCoreNLP-1.1.1.jar 100%[===================>]  26.14M  --.-KB/s    in 0.1s    

2023-05-25 02:51:12 (200 MB/s) - ‘VnCoreNLP-1.1.1.jar’ saved [27412575/27412575]

--2023-05-25 02:51:12--  https://raw.githubusercontent.com/vncorenlp/VnCoreNLP/master/models/wordsegmenter/vi-vocab
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.109.133, 185.199.111.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting respon

In [45]:
!git clone https://github.com/duyvuleo/VNTC.git
!ls VNTC/Data/10Topics/Ver1.1

fatal: destination path 'VNTC' already exists and is not an empty directory.
Stats.txt  Test_Full.rar  Train_Full.rar


In [13]:
!unrar x VNTC/Data/10Topics/Ver1.1/Train_Full.rar
!unrar x VNTC/Data/10Topics/Ver1.1/Test_Full.rar

Kết quả truyền trực tuyến bị cắt bớt đến 5000 dòng cuối.
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6148).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6210).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6079).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6083).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6086).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6088).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6312).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6092).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6094).txt                        89%  OK 
Extracting  Test_Full/Van hoa/VH_VNE_T_ (6101).txt                        89%  OK 
Extracting  Test_Full/V

In [46]:
import glob2
from tqdm import tqdm
from vncorenlp import VnCoreNLP
from fairseq.data import Dictionary
from fairseq.data.encoders.fastbpe import fastBPE
import numpy as np
import torch

rdrsegmenter = VnCoreNLP("vncorenlp/VnCoreNLP-1.1.1.jar", annotators="wseg", max_heap_size="-Xmx500m")

train_path = 'Train_Full/*/*.txt'
test_path = 'Test_Full/*/*.txt'

def read_txt(path):
  with open(path, 'r', encoding='utf-16') as f:
    data = f.read()
  return data

def get_stopwords_list():
  with open('vietnamese-stopwords-dash.txt', 'r', encoding='utf-8') as f:
    lines = f.readlines()
    stop_set = set(m.strip() for m in lines)
    return list(frozenset(stop_set))

stopwords = get_stopwords_list()
cnt = 0

def remove_stopwords(texts):
  global cnt
  tmp = texts.split(' ')
  for sw in stopwords:
    if sw in tmp:
      cnt += 1
      tmp.remove(sw)
  return ' '.join(tmp)

def remove_stopwords2(words_arr):
  global cnt
  for words in words_arr:
    for sw in stopwords:
      if sw in words:
        cnt += 1
        words.remove(sw)
  return words_arr

max_seq_len = 256
window_size = max_seq_len
def convert_line(content, vocab, bpe):
  real_len = max_seq_len - 2
  outputs = []
  cls_id = 0
  eos_id = 2
  pad_id = 1

  _content = ' '.join([' '.join(sent) for sent in rdrsegmenter.tokenize(content)])
  # t = rdrsegmenter.tokenize(content)
  # print('t original:', t)
  # t = remove_stopwords2(t)
  # print('t after removing stopwords', t)
  # _content = remove_stopwords(_content)
  # sentences = _content.split('.')
  # lines = []
  # text = []
  # for s in t:
  #   print('len(text)', len(text))
  #   print('len(s) + len(text)', len(s) + len(text))
  #   if len(text) == 0 or (len(s) + len(text) <= real_len):
  #     text = np.concatenate([text, s])
  #   else:
  #     l = ' '.join(text)
  #     print(l)
  #     # l = remove_stopwords(l)
  #     lines.append(l)
  #     text = np.concatenate([[], s])
  
  # print('lines c', lines)
  # for i, l in enumerate(lines):  
  subwords = '<s> ' + bpe.encode(_content) + ' </s>'
  input_ids = vocab.encode_line(subwords, append_eos=False, add_if_not_exist=False).long().tolist()
  if len(input_ids) > max_seq_len:
    input_ids = input_ids[:max_seq_len]
    input_ids[-1] = eos_id
  else:
    input_ids = input_ids + [pad_id,]*(max_seq_len - len(input_ids))
  outputs.append(np.array(input_ids))
  return outputs

def make_data(path, vocab, bpe):
  texts = []
  labels = []
  for file_path in tqdm(glob2.glob(path)):
    try:
      content = read_txt(file_path)
      lines = convert_line(content, vocab, bpe)
      label = file_path.split('/')[1]
      texts = np.concatenate([texts, lines])
      labels = np.concatenate([labels, [label,]*len(lines)])
      print(texts)
      print(labels)
    except:
      print('error')
      next
  return texts, labels

In [47]:
content = '''
Ngày xuân thưởng thức bánh Tổ Hội An.
Nhiều địa phương ở Quảng Nam đều làm bánh này. Nhưng ngày trước, ai đi mua, sắm Tết cũng thích ghé chợ Phố mua cho được bánh tổ Hội An. Trước cúng ông bà, sau dùng để thưởng thức. 
Bà Nguyễn Thị Hường, Tết Bính Tuất này tròn 68 tuổi, ở khối phố 2, phường Cẩm Phô Hội An tự hào đây là món đặc sản gia truyền mà gia đình bà ba đời nay vẫn còn giữ nghề. Bà Hường tỉ mẩn vừa làm, vừa giảng giải: 
Bánh tổ là loại bánh làm bằng bột nếp và đường. Nếp phải chọn loại tốt, dẻo và thơm, phơi thật khô rồi đem giả (ngày nay thì xay) thành bột. Đường làm bánh tổ là đường bát Quảng Nam (loại đường nấu thủ công, đổ vào bát ăn cơm, chờ cho cứng lại mới đem dùng (nên có tên là đường bát ). 
Đường đem "thắng" cho thật kỹ Đường đem "thắng" cho thật kỹ, loại bỏ hết tạp chất và thêm ít gừng tươi cho thêm hương vị, nếu khách thích đặt làm bánh tổ màu trắng thì làm bằng đường kết tinh. Bột, đường sau đó được trộn với nhau và được nhồi, đánh (bà Hường gọi là "giố" bột) thật kỹ. 
Điểm quan trọng ở khâu này (mang tính quyết định đối với việc làm ra cái bánh tổ và trở thành kỹ thuật riêng của người Hội An là "lấy trùng". Tức là tính toán để bột đường khi đã trở thành bánh thì không bị đặc, không bị nhão (bánh hỏng). 
Trộn bột, đường với nhau và nhồi 
Đổ nguyên liệu vào một cái "đài" làm bằng lá chuối tươi 
Thứ nguyên liệu sền sệt này bây giờ được đổ vào một cái "đài" đã được làm bằng lá chuối tươi và dùng tăm tre vót nhọn ghim kín mép lá. Chiếc "đài" bằng lá chuối bây giờ lại được đặt vào một rọ (khuôn) đan bằng tre để giữ cho bánh không bị biến dạng. 
Các khuôn bánh sau đó được đặt lên một tấm vỉ tre. Cứ 12 bánh xếp một vỉ. Lần lượt xếp từng vĩ vào nồi. Tất cả 6 vĩ/ nồi, cách nhau bởi hai thanh gỗ đan chéo nhau. Bánh tổ đem hấp liên tục trong 3 giờ thì chín, vớt ra để nguội. Nếu khách có nhu cầu bỏ mè thì khi vớt ra, người chủ nhanh tay rắc mè đã rang chín, làm sạch vỏ lên mặt bánh cho dính lại. 
Đặt khuôn bánh lên một tấm vỉ tre và từng vĩ vào nồi 
Cuối cùng, đem bánh ra phơi nắng độ hai hôm cho bốc hơi lượng nước ít ỏi trong bánh. Rồi đem ra chợ bày bán hoặc bỏ sĩ cho bạn hàng. 
Đặc trưng của bánh tổ Hội An là vừa dai vừa dẻo, hương vị đậm đà thơm ngon, có thể để được lâu. Khi ăn, có người thích cứ lấy cái bánh tổ xắt ra từng miếng và ăn ngay. Có người lại thích nướng trên bếp than hồng cho mềm đi rồi mới ăn. Nhưng cách nhiều người thích nhất là xắt miếng và chiên với dầu phộng. 
Miếng bánh tổ béo ngậy, thơm lừng mùi đường, mùi gừng, mùi nếp. Bỏ miếng bánh tổ vào miệng, nhai và nuốt đến đâu cảm giác vị ngọt, thanh lan toả đến đó. Đậm đà lắm. Thú vị lắm. Có lẽ cũng bởi thế mà dù bị bao nhiêu loại bánh kẹo công nghiệp lấn át nhưng bánh tổ Hội An vẫn sống mãi cùng quê hương mỗi độ Xuân về, Tết đến. 
'''

vocab = Dictionary()
vocab.add_from_file("PhoBERT_base_fairseq/dict.txt")
class BPE():
  bpe_codes = 'PhoBERT_base_fairseq/bpe.codes'
args = BPE()
bpe = fastBPE(args)

texts = convert_line(content, vocab, bpe)

In [48]:
from fairseq.models.roberta import RobertaModel

phoBERT = RobertaModel.from_pretrained('PhoBERT_base_fairseq', checkpoint='model.pt')
phoBERT.bpe = bpe


In [49]:
print('cnt:', cnt)
print('texts size', len(texts))
print('0:', texts[0])
phoBERT.decode(torch.tensor(texts[0]))

cnt: 0
texts size 1
0: [    0   319  2184  2483  1157 13438  4000     5   449   296    25  1831
   131    47  1157    23     5   293    43    71     4   277    57   188
     4  7032   708    32   543  4136   932  4670   188    13    11 56220
  3988  4000     5   347  4245  3604     4    53   175    24  2483     5
   680  1848  7407     4   708 31504 12190    23  2533  4277   126     4
    25 41078    76     4   557  5953  2474 12115  4000  2248    97     8
   526  4184 17820    64   161 33474   733   136    74    50   300   657
     5   680  7407 29544   164    47     4   164 34188    27 10121 12641
  3988     8   166  1157    47   121  1887  3962     6   109     5 33876
    41   437   166   167     4  8267     6  2023     4  4124   520  1427
   182   873  1163    20  2709    54  5158    19   302  1887     5  2080
    47 56220  3988     8   109  2838  1831    20   166   109  1872  4794
     4   888    33  2838   203  1412     4   971    13  2951    44    60
   873   175    20    77    

'Ngày xuân thưởng_thức bánh Tổ_Hội_An . Nhiều địa_phương ở Quảng_Nam đều làm bánh này . Nhưng ngày trước , ai đi mua , sắm Tết cũng thích ghé chợ Phố mua cho được bánh_tổ Hội_An . Trước cúng ông_bà , sau dùng để thưởng_thức . Bà Nguyễn_Thị_Hường , Tết Bính_Tuất này tròn 68 tuổi , ở khối_phố 2 , phường Cẩm_Phô_Hội_An tự_hào đây là món đặc_sản gia_truyền mà gia_đình bà_ba đời nay vẫn còn giữ nghề . Bà Hường tỉ_mẩn vừa làm , vừa giảng_giải : Bánh_tổ là loại bánh làm bằng bột nếp và đường . Nếp phải chọn loại tốt , dẻo và thơm , phơi thật khô rồi đem giả ( ngày_nay thì xay ) thành bột . Đường làm bánh_tổ là đường bát Quảng_Nam ( loại đường nấu thủ_công , đổ vào bát ăn cơm , chờ cho cứng lại mới đem dùng ( nên có tên là đường bát ) . Đường đem " thắng " cho thật kỹ Đường đem " thắng " cho thật kỹ , loại_bỏ hết tạp_chất và thêm ít gừng tươi cho thêm hương_vị , nếu khách thích đặt làm bánh_tổ màu trắng thì làm bằng đường kết_tinh . Bột , đường sau đó được trộn với nhau và được nhồi , đánh ( b

In [50]:
# Save data
import pickle

def _save_pkl(path, obj):
  with open(path, 'wb') as f:
    pickle.dump(obj, f)

def _load_pkl(path):
  with open(path, 'rb') as f:
    obj = pickle.load(f)
  return obj

In [ ]:
text_train, label_train = make_data(train_path, vocab, bpe)
text_test, label_test = make_data(test_path, vocab, bpe)
print(text_train[:2])
_save_pkl('text_train.pkl', text_train)
_save_pkl('label_train.pkl', label_train)
_save_pkl('text_test.pkl', text_test)
_save_pkl('label_test.pkl', label_test)

print('train size:', len(text_train))
print('test size:', len(text_test))

  0%|          | 6/33759 [00:00<21:56, 25.64it/s]

error
error
error
error
error
error


  0%|          | 9/33759 [00:00<24:37, 22.85it/s]

error
error
error


  0%|          | 12/33759 [00:00<29:17, 19.20it/s]

error
error
error
error


  0%|          | 18/33759 [00:00<30:39, 18.35it/s]

error
error
error
error
error


  0%|          | 20/33759 [00:01<34:46, 16.17it/s]

error
error
error


  0%|          | 22/33759 [00:01<35:27, 15.86it/s]

error
error


  0%|          | 24/33759 [00:01<41:15, 13.63it/s]

error
error


  0%|          | 26/33759 [00:01<59:35,  9.43it/s]

error


  0%|          | 28/33759 [00:02<1:07:30,  8.33it/s]

error
error


  0%|          | 31/33759 [00:02<1:02:50,  8.94it/s]

error
error
error


  0%|          | 33/33759 [00:02<1:14:00,  7.59it/s]

error
error


  0%|          | 35/33759 [00:02<1:05:44,  8.55it/s]

error
error
error


  0%|          | 37/33759 [00:03<53:11, 10.57it/s]  

error


  0%|          | 40/33759 [00:03<1:09:21,  8.10it/s]

error
error
error


  0%|          | 43/33759 [00:03<51:54, 10.83it/s]  

error
error
error
error


  0%|          | 47/33759 [00:04<48:26, 11.60it/s]

error
error
error
error


  0%|          | 52/33759 [00:04<38:34, 14.57it/s]

error
error
error
error


  0%|          | 57/33759 [00:04<32:57, 17.05it/s]

error
error
error
error
error


  0%|          | 63/33759 [00:04<30:52, 18.19it/s]

error
error
error
error
error
error
error


  0%|          | 66/33759 [00:05<30:35, 18.35it/s]

error
error
error
error


  0%|          | 72/33759 [00:05<30:21, 18.50it/s]

error
error
error
error


  0%|          | 76/33759 [00:05<30:28, 18.42it/s]

error
error
error
error
error


  0%|          | 78/33759 [00:05<35:31, 15.80it/s]

error
error


  0%|          | 82/33759 [00:06<42:58, 13.06it/s]

error
error
error
error


  0%|          | 86/33759 [00:06<44:11, 12.70it/s]

error
error
error


  0%|          | 89/33759 [00:06<37:33, 14.94it/s]

error
error
error


  0%|          | 91/33759 [00:06<41:51, 13.41it/s]

error
error
error


  0%|          | 95/33759 [00:07<47:21, 11.85it/s]

error
error
error


  0%|          | 97/33759 [00:07<49:11, 11.41it/s]

error
error
error


  0%|          | 101/33759 [00:07<43:42, 12.83it/s]

error
error
error
error
error


  0%|          | 106/33759 [00:08<39:06, 14.34it/s]

error
error
error
error


  0%|          | 111/33759 [00:08<33:34, 16.70it/s]

error
error
error
error
error


  0%|          | 117/33759 [00:08<27:37, 20.29it/s]

error
error
error
error
error


  0%|          | 124/33759 [00:08<22:56, 24.43it/s]

error
error
error
error
error
error
error


  0%|          | 127/33759 [00:08<24:04, 23.28it/s]

error
error
error
error
error
error


  0%|          | 136/33759 [00:09<19:05, 29.35it/s]

error
error
error
error
error
error
error
error


  0%|          | 144/33759 [00:09<17:43, 31.61it/s]

error
error
error
error
error
error
error
error


  0%|          | 152/33759 [00:09<16:48, 33.31it/s]

error
error
error
error
error
error
error
error


  0%|          | 160/33759 [00:09<18:18, 30.60it/s]

error
error
error
error
error
error


  0%|          | 164/33759 [00:10<18:48, 29.77it/s]

error
error
error
error
error
error
error
error


  1%|          | 173/33759 [00:10<19:26, 28.78it/s]

error
error
error
error
error


  1%|          | 178/33759 [00:10<18:16, 30.62it/s]

error
error
error
error
error
error
error
error
error


  1%|          | 184/33759 [00:10<16:05, 34.77it/s]

error
error
error
error
error


  1%|          | 192/33759 [00:10<18:25, 30.36it/s]

error
error
error
error
error
error
error


  1%|          | 200/33759 [00:11<17:29, 31.98it/s]

error
error
error
error
error
error
error


  1%|          | 204/33759 [00:11<22:00, 25.41it/s]

error
error
error
error


  1%|          | 211/33759 [00:11<19:33, 28.59it/s]

error
error
error
error
error
error


  1%|          | 215/33759 [00:11<23:05, 24.21it/s]

error
error
error
error
error
error
error


  1%|          | 222/33759 [00:12<22:05, 25.29it/s]

error
error
error
error
error
error
error


  1%|          | 230/33759 [00:12<20:03, 27.86it/s]

error
error
error
error
error
error
error


  1%|          | 233/33759 [00:12<20:37, 27.10it/s]

error
error


  1%|          | 236/33759 [00:12<31:12, 17.90it/s]

error
error
error
error


  1%|          | 242/33759 [00:13<30:21, 18.40it/s]

error
error
error
error


  1%|          | 245/33759 [00:13<30:09, 18.52it/s]

error
error
error
error
error


  1%|          | 252/33759 [00:13<28:23, 19.67it/s]

error
error
error
error
error


  1%|          | 255/33759 [00:13<31:11, 17.90it/s]

error
error
error
error


  1%|          | 262/33759 [00:14<28:21, 19.68it/s]

error
error
error
error
error
error


  1%|          | 265/33759 [00:14<27:00, 20.67it/s]

error
error
error
error
error
error


  1%|          | 272/33759 [00:14<28:55, 19.30it/s]

error
error
error
error
error

  1%|          | 276/33759 [00:14<26:39, 20.94it/s]


error
error
error
error
error


  1%|          | 283/33759 [00:15<22:41, 24.59it/s]

error
error
error
error
error
error


  1%|          | 286/33759 [00:15<29:51, 18.68it/s]

error
error
error


  1%|          | 292/33759 [00:15<29:14, 19.08it/s]

error
error
error
error
error
error


  1%|          | 296/33759 [00:15<28:04, 19.86it/s]

error
error
error
error
error
error


  1%|          | 301/33759 [00:16<24:38, 22.63it/s]

error
error
error
error
error


  1%|          | 308/33759 [00:16<22:54, 24.34it/s]

error
error
error
error
error
error


  1%|          | 315/33759 [00:16<20:30, 27.18it/s]

error
error
error
error
error
error


  1%|          | 325/33759 [00:16<15:54, 35.02it/s]

error
error
error
error
error
error
error
error
error
error


  1%|          | 334/33759 [00:17<14:39, 38.01it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  1%|          | 349/33759 [00:17<11:10, 49.85it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  1%|          | 362/33759 [00:17<10:26, 53.28it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  1%|          | 374/33759 [00:17<10:36, 52.42it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  1%|          | 387/33759 [00:18<10:01, 55.51it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error


  1%|          | 393/33759 [00:18<09:56, 55.91it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  1%|          | 406/33759 [00:18<12:16, 45.28it/s]

error
error
error
error
error
error
error


  1%|          | 411/33759 [00:18<13:42, 40.53it/s]

error
error
error
error
error
error
error


  1%|          | 416/33759 [00:18<13:57, 39.83it/s]

error
error
error
error
error
error
error


  1%|          | 421/33759 [00:19<18:10, 30.56it/s]

error
error
error
error


  1%|▏         | 428/33759 [00:19<21:50, 25.43it/s]

error
error
error
error


  1%|▏         | 431/33759 [00:19<24:44, 22.44it/s]

error
error
error
error


  1%|▏         | 434/33759 [00:19<33:05, 16.79it/s]

error
error
error


  1%|▏         | 439/33759 [00:20<32:11, 17.25it/s]

error
error
error
error
error
error
error


  1%|▏         | 447/33759 [00:20<24:04, 23.05it/s]

error
error
error
error
error
error


  1%|▏         | 453/33759 [00:20<26:10, 21.20it/s]

error
error
error
error
error


  1%|▏         | 457/33759 [00:20<26:05, 21.27it/s]

error
error
error
error


  1%|▏         | 463/33759 [00:21<26:32, 20.91it/s]

error
error
error
error
error
error


  1%|▏         | 466/33759 [00:21<27:24, 20.24it/s]

error
error
error
error
error


  1%|▏         | 469/33759 [00:21<26:41, 20.79it/s]

error
error


  1%|▏         | 477/33759 [00:21<25:03, 22.14it/s]

error
error
error
error
error
error
error


  1%|▏         | 480/33759 [00:22<25:58, 21.35it/s]

error
error
error
error
error


  1%|▏         | 488/33759 [00:22<19:19, 28.70it/s]

error
error
error
error
error
error
error
error
error


  1%|▏         | 496/33759 [00:22<17:31, 31.63it/s]

error
error
error
error
error
error
error
error
error


  1%|▏         | 502/33759 [00:22<14:22, 38.57it/s]

error
error
error
error
error
error


  2%|▏         | 511/33759 [00:22<16:27, 33.66it/s]

error
error
error
error
error
error
error
error
error


  2%|▏         | 520/33759 [00:23<17:45, 31.18it/s]

error
error
error
error
error
error


  2%|▏         | 528/33759 [00:23<17:04, 32.43it/s]

error
error
error
error
error
error
error
error


  2%|▏         | 536/33759 [00:23<16:26, 33.67it/s]

error
error
error
error
error
error
error


  2%|▏         | 546/33759 [00:23<13:52, 39.90it/s]

error
error
error
error
error
error
error
error
error
error


  2%|▏         | 551/33759 [00:24<14:10, 39.05it/s]

error
error
error
error
error
error


  2%|▏         | 559/33759 [00:24<17:34, 31.48it/s]

error
error
error
error
error
error
error


  2%|▏         | 563/33759 [00:24<17:23, 31.82it/s]

error
error
error
error
error
error


  2%|▏         | 567/33759 [00:24<19:51, 27.85it/s]

error
error
error
error
error


  2%|▏         | 576/33759 [00:24<17:11, 32.16it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 590/33759 [00:25<13:18, 41.56it/s]

error
error
error
error
error
error
error
error
error


  2%|▏         | 601/33759 [00:25<13:26, 41.09it/s]

error
error
error
error
error
error
error
error
error


  2%|▏         | 606/33759 [00:25<13:30, 40.93it/s]

error
error
error
error
error
error
error
error


  2%|▏         | 616/33759 [00:25<12:52, 42.90it/s]

error
error
error
error
error
error
error
error


  2%|▏         | 626/33759 [00:26<13:59, 39.46it/s]

error
error
error
error
error
error
error
error
error
error


  2%|▏         | 638/33759 [00:26<12:12, 45.22it/s]

error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 645/33759 [00:26<10:57, 50.37it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 658/33759 [00:26<10:03, 54.82it/s]

error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 670/33759 [00:26<11:14, 49.08it/s]

error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 675/33759 [00:27<11:22, 48.46it/s]

error
error
error
error
error
error
error


  2%|▏         | 685/33759 [00:27<13:41, 40.28it/s]

error
error
error
error
error
error
error
error
error


  2%|▏         | 696/33759 [00:27<12:04, 45.60it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 710/33759 [00:27<10:10, 54.12it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 716/33759 [00:27<10:49, 50.88it/s]

error
error
error
error
error
error
error
error


  2%|▏         | 728/33759 [00:28<12:07, 45.38it/s]

error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 741/33759 [00:28<11:44, 46.87it/s]

error
error
error
error
error
error
error
error
error
error


  2%|▏         | 746/33759 [00:28<11:50, 46.45it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 760/33759 [00:28<11:15, 48.87it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 774/33759 [00:29<10:57, 50.13it/s]

error
error
error
error
error
error
error
error
error


  2%|▏         | 787/33759 [00:29<10:39, 51.60it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 800/33759 [00:29<10:12, 53.81it/s]

error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 806/33759 [00:29<12:37, 43.52it/s]

error
error
error
error
error
error


  2%|▏         | 811/33759 [00:29<13:44, 39.96it/s]

error
error
error
error
error
error
error
error


  2%|▏         | 822/33759 [00:30<12:11, 45.05it/s]

error
error
error
error
error
error
error
error
error
error
error


  2%|▏         | 833/33759 [00:30<11:22, 48.21it/s]

error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 845/33759 [00:30<10:58, 49.98it/s]

error
error
error
error
error
error
error
error
error
error


  3%|▎         | 857/33759 [00:30<10:58, 49.96it/s]

error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 863/33759 [00:31<12:39, 43.33it/s]

error
error
error
error
error
error
error
error
error


  3%|▎         | 875/33759 [00:31<11:23, 48.13it/s]

error
error
error
error
error
error
error
error
error
error


  3%|▎         | 880/33759 [00:31<12:28, 43.94it/s]

error
error
error
error
error
error
error
error


  3%|▎         | 892/33759 [00:31<11:32, 47.43it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 902/33759 [00:31<12:26, 44.04it/s]

error
error
error
error
error
error
error
error
error
error


  3%|▎         | 913/33759 [00:32<11:50, 46.24it/s]

error
error
error
error
error
error
error
error
error


  3%|▎         | 923/33759 [00:32<13:05, 41.80it/s]

error
error
error
error
error
error
error
error


  3%|▎         | 934/33759 [00:32<12:32, 43.64it/s]

error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 949/33759 [00:32<09:46, 55.93it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 956/33759 [00:32<09:16, 58.90it/s]

error
error
error
error
error
error
error
error
error


  3%|▎         | 968/33759 [00:33<11:09, 48.99it/s]

error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 981/33759 [00:33<10:19, 52.90it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 987/33759 [00:33<10:18, 53.01it/s]

error
error
error
error
error
error
error
error
error
error


  3%|▎         | 1000/33759 [00:33<10:27, 52.22it/s]

error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 1006/33759 [00:34<11:32, 47.29it/s]

error
error
error
error
error
error
error
error
error


  3%|▎         | 1017/33759 [00:34<11:50, 46.05it/s]

error
error
error
error
error
error
error
error
error


  3%|▎         | 1027/33759 [00:34<12:54, 42.29it/s]

error
error
error
error
error
error
error
error


  3%|▎         | 1039/33759 [00:34<10:56, 49.83it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 1045/33759 [00:34<12:11, 44.75it/s]

error
error
error
error
error
error
error
error
error
error
error


  3%|▎         | 1057/33759 [00:35<11:53, 45.85it/s]

error
error
error
error
error
error
error
error
error


  3%|▎         | 1067/33759 [00:35<13:33, 40.20it/s]

error
error
error
error
error
error
error


  3%|▎         | 1072/33759 [00:35<14:04, 38.70it/s]

error
error
error
error
error
error
error


  3%|▎         | 1077/33759 [00:35<15:56, 34.19it/s]

error
error
error
error
error


  3%|▎         | 1086/33759 [00:36<16:58, 32.07it/s]

error
error
error
error
error
error
error
error
error
error


  3%|▎         | 1094/33759 [00:36<18:16, 29.78it/s]

error
error
error
error
error
error


  3%|▎         | 1104/33759 [00:36<15:25, 35.28it/s]

error
error
error
error
error
error
error
error
error


  3%|▎         | 1114/33759 [00:36<13:26, 40.45it/s]

error
error
error
error
error
error
error
error
error
error
error

  3%|▎         | 1124/33759 [00:37<12:30, 43.47it/s]


error
error
error
error
error
error
error
error
error


  3%|▎         | 1129/33759 [00:37<13:29, 40.30it/s]

error
error
error
error
error
error
error
error


  3%|▎         | 1139/33759 [00:37<15:32, 34.98it/s]

error
error
error
error
error
error
error


  3%|▎         | 1143/33759 [00:37<15:36, 34.85it/s]

error
error
error
error
error
error
error


  3%|▎         | 1151/33759 [00:37<15:13, 35.69it/s]

error
error
error
error
error
error
error
error


  3%|▎         | 1159/33759 [00:38<14:44, 36.88it/s]

error
error
error
error
error
error
error


  3%|▎         | 1163/33759 [00:38<16:16, 33.39it/s]

error
error
error
error
error


  3%|▎         | 1172/33759 [00:38<17:16, 31.42it/s]

error
error
error
error
error
error
error
error
error
error


  3%|▎         | 1181/33759 [00:38<16:53, 32.13it/s]

error
error
error
error
error
error


  4%|▎         | 1185/33759 [00:38<16:38, 32.64it/s]

error
error
error
error
error


  4%|▎         | 1193/33759 [00:39<17:10, 31.59it/s]

error
error
error
error
error
error
error
error
error


  4%|▎         | 1202/33759 [00:39<14:32, 37.33it/s]

error
error
error
error
error
error
error
error


  4%|▎         | 1210/33759 [00:39<15:36, 34.76it/s]

error
error
error
error
error
error
error
error


  4%|▎         | 1218/33759 [00:39<16:51, 32.18it/s]

error
error
error
error
error
error
error


  4%|▎         | 1228/33759 [00:40<13:52, 39.06it/s]

error
error
error
error
error
error
error
error
error
error
error


  4%|▎         | 1238/33759 [00:40<13:16, 40.85it/s]

error
error
error
error
error
error
error
error


  4%|▎         | 1243/33759 [00:40<13:57, 38.81it/s]

error
error
error
error
error
error
error
error
error

  4%|▎         | 1252/33759 [00:40<13:50, 39.14it/s]


error
error
error
error
error
error
error
error


  4%|▎         | 1261/33759 [00:40<14:48, 36.60it/s]

error
error
error
error
error
error
error
error


  4%|▎         | 1265/33759 [00:41<22:10, 24.42it/s]

error
error
error
error


  4%|▍         | 1271/33759 [00:41<29:50, 18.14it/s]

error
error
error
error
error
error
error


  4%|▍         | 1284/33759 [00:41<16:45, 32.31it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1294/33759 [00:42<14:37, 37.00it/s]

error
error
error
error
error
error
error
error
error


  4%|▍         | 1305/33759 [00:42<12:27, 43.43it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1317/33759 [00:42<11:24, 47.41it/s]

error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1328/33759 [00:42<11:25, 47.32it/s]

error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1336/33759 [00:43<11:50, 45.65it/s]

error
error
error
error
error
error
error
error
error


  4%|▍         | 1349/33759 [00:43<10:42, 50.42it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1355/33759 [00:43<10:55, 49.45it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1372/33759 [00:43<08:50, 61.04it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1388/33759 [00:43<07:42, 70.01it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1396/33759 [00:44<08:27, 63.76it/s]

error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1409/33759 [00:44<10:27, 51.52it/s]

error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1421/33759 [00:44<10:15, 52.58it/s]

error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1434/33759 [00:44<09:33, 56.34it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1446/33759 [00:45<11:05, 48.55it/s]

error
error
error
error
error
error
error
error
error


  4%|▍         | 1452/33759 [00:45<12:52, 41.80it/s]

error
error
error
error
error
error


  4%|▍         | 1457/33759 [00:45<13:59, 38.49it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1470/33759 [00:45<11:58, 44.97it/s]

error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1480/33759 [00:45<11:32, 46.60it/s]

error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1491/33759 [00:46<11:39, 46.14it/s]

error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1504/33759 [00:46<10:24, 51.63it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  4%|▍         | 1510/33759 [00:46<10:01, 53.60it/s]

error
error
error
error
error
error
error


  5%|▍         | 1521/33759 [00:46<12:26, 43.17it/s]

error
error
error
error
error
error
error
error
error


  5%|▍         | 1533/33759 [00:46<11:43, 45.82it/s]

error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1545/33759 [00:47<10:51, 49.45it/s]

error
error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1560/33759 [00:47<09:49, 54.66it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1566/33759 [00:47<09:57, 53.86it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1582/33759 [00:47<08:33, 62.65it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1595/33759 [00:48<09:25, 56.91it/s]

error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1607/33759 [00:48<10:08, 52.84it/s]

error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1613/33759 [00:48<10:47, 49.68it/s]

error
error
error
error
error
error
error
error


  5%|▍         | 1626/33759 [00:48<10:58, 48.77it/s]

error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1641/33759 [00:48<09:58, 53.67it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1647/33759 [00:49<10:39, 50.18it/s]

error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1660/33759 [00:49<10:57, 48.82it/s]

error
error
error
error
error
error
error
error
error


  5%|▍         | 1665/33759 [00:49<11:21, 47.10it/s]

error
error
error
error
error
error
error
error


  5%|▍         | 1676/33759 [00:49<11:08, 48.00it/s]

error
error
error
error
error
error
error
error
error
error
error
error
error


  5%|▍         | 1683/33759 [00:49<10:06, 52.86it/s]

In [23]:
text_train_loaded = _load_pkl('text_train.pkl')
print('load:', len(text_train_loaded), text_train_loaded[:2])

load: 0 []


In [24]:
from fairseq.data import Dictionary
from fairseq.data.encoders.fastbpe import fastBPE
import numpy as np
import torch

# Tokenize noi dung
max_seq_len = 256

def convert_lines(lines, vocab, bpe):
  '''
  lines: list các văn bản input
  vocab: từ điển dùng để encoding subwords
  bpe:
  '''
  # khoi tao ma tran output
  outputs = np.zeros((len(lines), max_seq_len), dtype=np.int32) # shape (number_lines, max_seq_len)
  # Index cac token cls (đầu câu), eos (cuối câu), padding (padding token)
  cls_id = 0
  eos_id = 2
  pad_id = 1

  for i, row in tqdm(enumerate(lines), total=len(lines)):
    # Mã hoá subwords theo bpe
    subwords = bpe.encode('<s> ' + row + ' </s>')
    input_ids = vocab.encode_line(subwords, append_eos=False, add_if_not_exist=False).long().tolist()
    # truncate input nếu độ dài vượt quá max_seq_len
    if len(input_ids) > max_seq_len:
      input_ids = input_ids[:max_seq_len]
      input_ids[-1] = eos_id
    else:
      # Padding nếu độ dài câu chưa bằng max_seq_len
      input_ids += [pad_id,]*(max_seq_len - len(input_ids))
    outputs[i, :] = np.array(input_ids)
  return outputs


# Load the dictionary  
vocab = Dictionary()
vocab.add_from_file("PhoBERT_base_fairseq/dict.txt")

class BPE():
  bpe_codes = 'PhoBERT_base_fairseq/bpe.codes'
args = BPE()
bpe = fastBPE(args)

# Test encode lines
# lines = ['Học_sinh được nghỉ học bắt dầu từ tháng 3 để tránh dịch covid-19', 'số lượng ca nhiễm bệnh đã giảm bắt đầu từ tháng 5 nhờ biện pháp mạnh tay']
lines = ['Ông đồ cuối_cùng trên đảo “ Vua_Tần cỡi cọp chơi tám cõi Trời xanh , kiếm quang sáng chói ... Tiên_nhân cầm đuốc khói nhẹ_nhàng Mắt say nhoà lệ theo cung đàn ” . Ngồi trên bãi biển lộng_gió , ông đồ Võ_Hiển_Đạt bất_ngờ đọc cho tôi nghe bài Tần vương ẩm tửu của Lý_Hạ , người được mệnh_danh là “ thi quỉ ” đời Đường , tiếng_tăm và tác_phẩm truyền lưu hậu_thế không thua_kém “ thi tiên ” Lý_Bạch , “ thi Phật ” Vương_Duy , “ thi thánh ” Đỗ_Phủ . Rồi ông tâm_sự : “ Tôi sinh ra nhằm lúc Nho_học mạt_vận , may_mắn được học dăm chữ thánh_hiền . Bây_giờ , thỉnh_thoảng tôi chỉ còn dùng mớ chữ_nghĩa một thời vang bóng này khi có bạn tâm_giao hoặc ai đó nhờ viết câu_đối xưa ” . Men theo con đường bãi biển rợp bóng dừa xanh , tôi về nhà ông ở thôn Tây , xã An_Vĩnh , huyện đảo Lý_Sơn ( Quảng_Ngãi ) . Thời_gian như dừng lại trong ngôi nhà cổ của ông đồ cuối_cùng ở hòn đảo này . Những bức hoành_phi , câu_đối được viết bằng mực_tàu , bút_lông nét chữ như rồng múa phượng bay treo từ hàng cột bờ hiên vào tận bàn_thờ tổ_tiên giữa gian nhà chính . Bộ sập gụ đã trải hơn trăm_năm sáng bóng_nước thời_gian . Đó là nơi ông thường ngồi_xếp_bằng để đọc sách và múa bút khi có người xin chữ . Trang_trọng nhất trong ngôi nhà cổ là chiếc tủ bằng gỗ lim lúc_nào cũng ấm mùi nhang trầm phảng_phất trên hàng linh_vị thờ các vị tổ_tiên . Bút tự của ông đồ Đạt có ở khắp_nơi trên đảo Lý_Sơn Vượt biển ra Lý_Sơn , tôi tình_cờ biết ông đồ Võ_Hiển_Đạt khi được chứng_kiến một buổi an_táng theo nghi_lễ mộ gió các ngư_dân bất_hạnh phải gửi xác dưới lòng đại_dương . Thi_hài trong quan_tài nhỏ chỉ là hình_nhân đất_sét , nhưng lễ_nghi vẫn nghiêm linh với lời tụng kinh , cờ phướn cầu_siêu những người đã ra đi mãi_mãi không về . Anh cán_bộ văn_phòng uỷ_ban huyện cho biết : “ Gần nửa thế_kỷ qua , đảo này chỉ có một thầy_đồ duy_nhất và cũng có_lẽ là thầy_đồ cuối_cùng . Không_chỉ ở các đình chùa mà tất_cả chữ Nho trong bất_cứ ngôi nhà nào trên đảo đều do một tay ông viết ... ” . Ông kể hồi còn nhỏ vùng này chỉ có một trường tiểu_học dạy tiếng Pháp ; ai muốn học cao hơn phải ra Huế hoặc vào Qui_Nhơn . Cụ thân_sinh của ông là một đồ nho ẩn_dật nhưng phóng_khoáng , cho con đi học tiếng Pháp nhưng vẫn khuyên ông nên dùi_mài_kinh_sử với Nho_học . Đến bây_giờ , ông vẫn nhớ lời cha dạy : “ Chữ Pháp giúp con mở_mang văn_minh , văn_hoá . Nhưng chữ Nho mới đủ thâm_sâu để dạy con đạo làm người . Con_người sinh ra_đời không đem đến được gì và chết cũng chẳng mang theo được cái chi , cuối_cùng chỉ có cái đạo làm người là ở lại ” . Nghe lời cha , ông theo cả Nho_học lẫn Pháp ngữ nhưng lấy chữ Nho làm trọng . Ông may_mắn có nhiều người thầy khả_kính uyên_thâm cả chữ Nho lẫn chữ Pháp . Chín năm theo các thầy , thêm mười năm tự học từ sách_vở , đến năm 30 tuổi Võ_Hiển_Đạt đã trở_thành một người dạy chữ và cho chữ thánh_hiền : “ Đó cũng là lúc đất_nước giao_thời , hầu_hết lớp trẻ đã chuyển sang dùi_mài quốc_ngữ hoặc tiếng Pháp . Chữ_Nho dần_dần trở_thành linh tự trên các bàn_thờ , đình chùa và trong câu_đối của các bậc lớn_tuổi giáo_huấn lớp hậu_sinh gìn_giữ đạo_đức , gia_phong ... ” . Đến giờ ông vẫn còn gìn_giữ nguyên_vẹn bộ Tứ_Thư được in từ năm 1942 và nhiều đầu_sách cổ khác . Ngôi nhà cổ của ông đồ trong mùa thu_hoạch tỏi Không gặp thời , ông nối bước cha lui về ẩn_dật ở hòn đảo nhỏ giữa biển . Ban_ngày vừa làm nghề cá vừa trồng rau khoai kiếm sống thanh_nhàn , chiều_tối mở lớp gõ_đầu_trẻ . Tuy_nhiên , cho_dù kính_trọng chữ thánh_hiền nhưng nhiều gia_đình cũng không muốn con_em mình theo Nho_học chỉ còn một thời vang bóng , vừa khó học vừa không_thể xin được việc_làm . Thế_là lớp_học của ông cứ vắng dần , vắng dần cho đến lúc không_thể tồn_tại được nữa . Không gõ_đầu_trẻ nữa nhưng ông trở_thành một người cho chữ thánh_hiền nổi_tiếng ở Lý_Sơn dù tuổi mới ngoài tam thập . Chữ của ông trầm_mặc , u_buồn ở Âm linh tự , nơi thờ_cúng đội lính Hoàng_Sa kiêu_hùng thuở xưa và những người đã bỏ_mạng giữa biển khơi . Chữ của ông trang_nghiêm như lời gia_huấn trên liễn thờ các bậc tiền hiền . Chữ của ông cũng huyền_ảo , vi diệu trên bàn_thờ Nam_Hải tướng_quân , vị thần tôn_kính của ngư_dân quanh_năm treo mạng trên đầu_sóng_ngọn_gió . Người vui cũng tìm đến ông để xin mấy chữ mới treo trên bàn_thờ tổ_tiên trong ngày cưới_xin , cất nhà mới . Người buồn cũng nài ông vài chữ cho trọn đạo_nghĩa trong lễ tang ma . Lúc_nào ông cũng vui_vẻ viết tặng khi có người đến xin chữ , bởi theo ông đó không_chỉ là bổn_phận mà_còn là cái đạo của người theo Nho_học , chọn nhân_nghĩa để sống ở đời ... Đã sắp đến hồi về với tổ_tiên , ông nói không còn gì phải hối_tiếc nhưng chỉ buồn một điều là chưa có một người học_trò chân_truyền nào để gửi_gắm lại cái chữ , cái đạo của thánh_hiền , kể_cả con_cái . Con ông người bám đảo làm nghề biển , người vào Sài_Gòn bươn_chải mưu_sinh ... “ Nhiều lúc tôi đành ngẫm_nghĩ hình_như cái gì cũng có vận thịnh_suy của nó . Nho_học cũng vậy thôi ! ” - ông trầm_ngâm tâm_sự rồi mài mực , trải giấy , khai_bút tặng tôi một chữ “ Tâm ” làm hành_trang trở về với đất_liền ... Theo ông đồ Đạt , người cho chữ thánh_hiền ngày_nay tuy không còn khắt_khe như xưa nữa nhưng vẫn có một_số điều cần tôn_trọng , gìn_giữ để bút_lực có_thể đạt đến sự tinh_tế , đến cái hồn sâu thẳm của từng nét chữ . Ông không viết khi mưa_gió . Ông cũng không cầm bút khi trời mây âm_u , hay lúc tịch_dương ngả bóng ảm_đạm . Ông thường đề bút vào buổi sáng , thời_điểm mặt_trời bắt_đầu toả nắng ấm . Sau khi thắp nén hương trầm lên bàn_thờ và cạn mấy tách trà , ông mới chấm bút_lông vào mực . Và lúc nét bút bắt_đầu lướt trên giấy đỏ , tâm người viết thanh_tịnh như mặt_nước hồ thu , không vọng danh_lợi , không sợ không khinh , cũng không vui không buồn ... Tất_cả chỉ còn có chữ , và tâm người viết sẽ bộc_lộ trên cái thần , cái hồn của nét chữ .']
[x1] = convert_lines(lines, vocab, bpe)
print('\n')
print('x1 tensor encode: {}, shape: {}'.format(x1[:10], x1.size))
# print('x1 tensor decode: ', phoBERT.decode(torch.tensor(x1))[:103])


100%|██████████| 1/1 [00:00<00:00, 62.61it/s]



x1 tensor encode: [63117  1302   884   168   779   710    34   705   556  7481], shape: 256


In [25]:
# load data
X = convert_lines(text_train, vocab, bpe)

from sklearn.preprocessing import LabelEncoder
label_train = _load_pkl('label_train.pkl')
lb = LabelEncoder()
lb.fit(label_train)
y = lb.fit_transform(label_train)
print(lb.classes_)
print('Top 5 classes indices: ', y[:5])

[]
Top 5 classes indices:  []


In [ ]:
# Save dữ liệu
_save_pkl('PhoBERT_pretrain/X_train_both.pkl', X)
_save_pkl('PhoBERT_pretrain/y_train_both.pkl', y)
_save_pkl('PhoBERT_pretrain/labelEncoder1.pkl', lb)

In [34]:
# train
import time
import random
import os
import numpy as np
from tqdm import tqdm

from transformers import *
from transformers.modeling_utils import * 

import torch
from torch import nn

from sklearn.model_selection import StratifiedKFold, StratifiedShuffleSplit
from sklearn.metrics import accuracy_score, f1_score

# Khởi tạo argument
EPOCHS = 10
BATCH_SIZE = 16
ACCUMULATION_STEPS = 5
FOLD = 4
LR = 2e-5
LR_DC_STEP = 80
LR_DC = 0.1
CUR_DIR = os.path.dirname(os.getcwd())
DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
CKPT_PATH = 'model_ckpt_full_both'
SEED = 69

TRAIN_PATH = 'X_train_both.pkl'
LABEL_PATH = 'y_train_both.pkl'
DICT_PATH = 'PhoBERT_base_fairseq/dict.txt'
BPE_PATH = 'PhoBERT_base_fairseq/bpe.codes'
PRETRAINED_PATH = 'PhoBERT_base_fairseq'

def seed_everything(SEED):
  np.random.seed(SEED)
  torch.manual_seed(SEED)
  torch.cuda.manual_seed(SEED)
  torch.backends.cudnn.deterministic = True

seed_everything(SEED)

if not os.path.exists(CKPT_PATH):
  os.mkdir(CKPT_PATH)

In [29]:
X_train = _load_pkl(TRAIN_PATH)
y = _load_pkl(LABEL_PATH)
print(X_train.shape)
print(y[:5])

FileNotFoundError: ignored

In [ ]:
from fairseq.data.encoders.fastbpe import fastBPE
from fairseq.models.roberta import RobertaModel

# Load model
class BPE():
  bpe_codes = f'{PRETRAINED_PATH}/bpe.codes'
  
print('Loading model...')
args = BPE()
bpe = fastBPE(args)
model = RobertaModel.from_pretrained(PRETRAINED_PATH, checkpoint_file='model.pt')
model.bpe = bpe
model.eval()
# Add header cho classification với số lượng classes = 10
model.register_classification_head('new_task', num_classes=10)
model.to(DEVICE)
print('Model has loaded\n')
print(model)

Loading model...


1042301B [00:01, 782812.72B/s]
456318B [00:00, 508582.39B/s]


Model has loaded

RobertaHubInterface(
  (model): RobertaModel(
    (encoder): RobertaEncoder(
      (sentence_encoder): TransformerEncoder(
        (dropout_module): FairseqDropout()
        (embed_tokens): Embedding(64001, 768, padding_idx=1)
        (embed_positions): LearnedPositionalEmbedding(258, 768, padding_idx=1)
        (layernorm_embedding): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (layers): ModuleList(
          (0): TransformerEncoderLayerBase(
            (self_attn): MultiheadAttention(
              (dropout_module): FairseqDropout()
              (k_proj): Linear(in_features=768, out_features=768, bias=True)
              (v_proj): Linear(in_features=768, out_features=768, bias=True)
              (q_proj): Linear(in_features=768, out_features=768, bias=True)
              (out_proj): Linear(in_features=768, out_features=768, bias=True)
            )
            (self_attn_layer_norm): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
         

In [ ]:
tokens = 'Học_sinh được nghỉ học bắt đầu từ tháng 3 do ảnh hưởng của dịch covid-19'
token_idxs = model.encode(tokens)
model.to(DEVICE)
logprobs = model.predict('new_task', token_idxs)  # tensor([[-1.1050, -1.0672, -1.1245]], grad_fn=<LogSoftmaxBackward>)
logprobs

tensor([[-2.4380, -2.1180, -2.2356, -2.2369, -2.2857, -2.4181, -2.2478, -2.3526,
         -2.3118, -2.4298]], device='cuda:0', grad_fn=<LogSoftmaxBackward0>)

In [ ]:
# if torch.cuda.device_count():
#   print(f"Training using {torch.cuda.device_count()} gpus")
#   model = nn.DataParallel(model)

# Creating optimizer and lr schedulers
param_optimizer = list(model.named_parameters())
no_decay = ['bias', 'LayerNorm.bias', 'LayerNorm.weight']
optimizer_grouped_parameters = [
  {'params': [p for n, p in param_optimizer if not any(nd in n for nd in no_decay)], 'weight_decay': 0.01},
  {'params': [p for n, p in param_optimizer if any(nd in n for nd in no_decay)], 'weight_decay': 0.0}
]
num_train_optimization_steps = int(EPOCHS*len(X_train)/BATCH_SIZE/ACCUMULATION_STEPS)
optimizer = torch.optim.AdamW(optimizer_grouped_parameters, lr=LR)  # To reproduce BertAdam specific behavior set correct_bias=False
scheduler = get_linear_schedule_with_warmup(optimizer, num_warmup_steps=100, num_training_steps=num_train_optimization_steps)  # PyTorch scheduler
scheduler0 = get_constant_schedule(optimizer)  # PyTorch scheduler
criteria = nn.NLLLoss()
print('created optimizer, scheduler, loss function')

created optimizer, scheduler, loss function


In [ ]:
def evaluate(logits, targets):
  """
  Đánh giá model sử dụng accuracy và f1 scores.
  Args:
      logits (B,C): torch.LongTensor. giá trị predicted logit cho class output.
      targets (B): torch.LongTensor. actual target indices.
  Returns:
      acc (float): the accuracy score
      f1 (float): the f1 score
  """
  # Tính accuracy score và f1_score
  logits = logits.detach().cpu().numpy()    
  y_pred = np.argmax(logits, axis = 1)
  targets = targets.detach().cpu().numpy()
  f1 = f1_score(targets, y_pred, average='weighted')
  acc = accuracy_score(targets, y_pred)
  return acc, f1

def validate(valid_loader, model, device):
  model.eval()
  accs = []
  f1s = []
  losses = []
  with torch.no_grad():
    for x_batch, y_batch in valid_loader:
      x_batch = x_batch.to(device)
      y_batch = y_batch.to(device)
      outputs = model.predict('new_task', x_batch)
      loss = criteria(outputs, y_batch)
      logits = torch.exp(outputs)
      acc, f1 = evaluate(logits, y_batch)
      accs.append(acc)
      f1s.append(f1)
      losses.append(loss.detach().cpu().numpy())
  
  mean_acc = np.mean(accs)
  mean_f1 = np.mean(f1s)
  mean_loss = np.mean(losses)
  return mean_acc, mean_f1, mean_loss

def train_on_epoch(train_loader, model, optimizer, scheduler, epoch, num_epochs, criteria, device, log_aggr = 100):
  model.train()
  sum_epoch_loss = 0
  sum_acc = 0
  sum_f1 = 0
  start = time.time()
  train_size = len(train_loader)

  for i, (x_batch, y_batch) in enumerate(train_loader):
    x_batch = x_batch.to(device)
    y_batch = y_batch.to(device)
    # model.train()
    # Compute prediction and loss
    y_pred = model.predict('new_task', x_batch)
    logits = torch.exp(y_pred)
    acc, f1 = evaluate(logits, y_batch)
    loss = criteria(y_pred, y_batch)

    # Backpropagation
    loss.backward()
    if i % ACCUMULATION_STEPS == 0 or i == train_size- 1:
      optimizer.step()
      optimizer.zero_grad()
      scheduler.step()

    loss_val = loss.item()
    sum_epoch_loss += loss_val
    sum_acc += acc
    sum_f1 += f1

    # pbar.set_postfix(loss = loss_val)

    if i % log_aggr == 0 or i == train_size - 1:
      print(f'[TRAIN] epoch {epoch + 1}/{num_epochs}, \
        observation {i}/{train_size} batch, \
        loss: {loss:>.4f} (avg {(sum_epoch_loss / (i + 1)):>.4f}), \
        avg acc: {(sum_acc / (i + 1)):>.4f}, \
        avg f1: {(sum_f1 / (i + 1)):>.4f}, \
        ({(len(x_batch) / (time.time() - start)):>.2f} im/s)')
    start = time.time()

# splits = list(StratifiedKFold(n_splits=1, shuffle=True, random_state=123).split(X_train, y))
splits = list(StratifiedShuffleSplit(n_splits=1, random_state=123, test_size=0.1).split(X_train, y))
for fold, (train_idx, val_idx) in enumerate(splits):
  # if fold != FOLD:
  #   continue
  best_score = 0
  print("Training for fold {}".format(fold))

  # Create dataset
  train_dataset = torch.utils.data.TensorDataset(torch.tensor(X_train[train_idx],dtype=torch.long), torch.tensor(y[train_idx],dtype=torch.long))
  valid_dataset = torch.utils.data.TensorDataset(torch.tensor(X_train[val_idx],dtype=torch.long), torch.tensor(y[val_idx],dtype=torch.long))
  # train_dataset = torch.utils.data.TensorDataset(torch.tensor(X_train,dtype=torch.long), torch.tensor(y,dtype=torch.long))
  # valid_dataset = torch.utils.data.TensorDataset(torch.tensor(X_train,dtype=torch.long), torch.tensor(y,dtype=torch.long))
  print('train size:', len(train_dataset))
  print('valid size:', len(valid_dataset))
  # for child in model.children():
  #   for param in child.parameters():
  #     # if not param.requires_grad:
  #       # print('whoopsies')
  #     param.requires_grad = False

  frozen = True

  for epoch in tqdm(range(EPOCHS + 1)):
    
    if epoch > 0 and frozen:
      for child in model.children():
        for param in child.parameters():
          param.requires_grad = True
      frozen = False
      del scheduler0
      torch.cuda.empty_cache()

    # Create DataLoader
    train_loader = torch.utils.data.DataLoader(train_dataset, batch_size=BATCH_SIZE, shuffle=True)
    valid_loader = torch.utils.data.DataLoader(valid_dataset, batch_size=BATCH_SIZE, shuffle=False)

    # Train model on EPOCH
    print('\nEpoch: ', epoch + 1)
    optimizer.zero_grad()
    train_on_epoch(train_loader=train_loader,
                  model=model,
                  optimizer=optimizer, 
                  epoch=epoch, 
                  num_epochs=EPOCHS + 1, 
                  criteria=criteria, 
                  device=DEVICE, 
                  scheduler=(scheduler0 if frozen else scheduler))
    
    # Validate on validation set
    acc, f1, loss = validate(valid_loader, model, device=DEVICE)
    print('Epoch {} validation acc: {:.4f}, f1: {:.4f}, loss: {:.4f} \n'.format(epoch + 1, acc, f1, loss))
    if f1 >= best_score:
      torch.save(model.state_dict(), os.path.join(CKPT_PATH, f'model.bin'))
      best_score = f1



Training for fold 0
train size: 30383
valid size: 3376


  0%|          | 0/11 [00:00<?, ?it/s]


Epoch:  1
[TRAIN] epoch 1/11,         observation 0/1899 batch,         loss: 2.3239 (avg 2.3239),         avg acc: 0.1250,         avg f1: 0.1036,         (5.33 im/s)
[TRAIN] epoch 1/11,         observation 100/1899 batch,         loss: 1.8899 (avg 2.1175),         avg acc: 0.3311,         avg f1: 0.2629,         (21.53 im/s)
[TRAIN] epoch 1/11,         observation 200/1899 batch,         loss: 1.3744 (avg 1.8563),         avg acc: 0.5432,         avg f1: 0.4985,         (21.40 im/s)
[TRAIN] epoch 1/11,         observation 300/1899 batch,         loss: 0.6711 (avg 1.5882),         avg acc: 0.6437,         avg f1: 0.6099,         (21.40 im/s)
[TRAIN] epoch 1/11,         observation 400/1899 batch,         loss: 1.0998 (avg 1.3764),         avg acc: 0.6986,         avg f1: 0.6709,         (21.37 im/s)
[TRAIN] epoch 1/11,         observation 500/1899 batch,         loss: 0.4506 (avg 1.2202),         avg acc: 0.7338,         avg f1: 0.7115,         (21.44 im/s)
[TRAIN] epoch 1/11,       

  9%|▉         | 1/11 [23:09<3:51:30, 1389.01s/it]


Epoch:  2
[TRAIN] epoch 2/11,         observation 0/1899 batch,         loss: 0.2650 (avg 0.2650),         avg acc: 0.9375,         avg f1: 0.9417,         (21.83 im/s)
[TRAIN] epoch 2/11,         observation 100/1899 batch,         loss: 0.1968 (avg 0.2091),         avg acc: 0.9369,         avg f1: 0.9353,         (21.40 im/s)
[TRAIN] epoch 2/11,         observation 200/1899 batch,         loss: 0.1375 (avg 0.2027),         avg acc: 0.9391,         avg f1: 0.9372,         (21.76 im/s)
[TRAIN] epoch 2/11,         observation 300/1899 batch,         loss: 0.3141 (avg 0.2098),         avg acc: 0.9383,         avg f1: 0.9381,         (21.55 im/s)
[TRAIN] epoch 2/11,         observation 400/1899 batch,         loss: 0.0816 (avg 0.2045),         avg acc: 0.9389,         avg f1: 0.9383,         (22.00 im/s)
[TRAIN] epoch 2/11,         observation 500/1899 batch,         loss: 0.2244 (avg 0.2031),         avg acc: 0.9397,         avg f1: 0.9393,         (21.40 im/s)
[TRAIN] epoch 2/11,      

 18%|█▊        | 2/11 [46:15<3:28:06, 1387.35s/it]


Epoch:  3
[TRAIN] epoch 3/11,         observation 0/1899 batch,         loss: 0.1182 (avg 0.1182),         avg acc: 1.0000,         avg f1: 1.0000,         (20.57 im/s)
[TRAIN] epoch 3/11,         observation 100/1899 batch,         loss: 0.4060 (avg 0.1365),         avg acc: 0.9604,         avg f1: 0.9603,         (21.56 im/s)
[TRAIN] epoch 3/11,         observation 200/1899 batch,         loss: 0.0426 (avg 0.1368),         avg acc: 0.9621,         avg f1: 0.9615,         (21.53 im/s)
[TRAIN] epoch 3/11,         observation 300/1899 batch,         loss: 0.0172 (avg 0.1360),         avg acc: 0.9603,         avg f1: 0.9603,         (21.49 im/s)
[TRAIN] epoch 3/11,         observation 400/1899 batch,         loss: 0.1109 (avg 0.1346),         avg acc: 0.9613,         avg f1: 0.9615,         (21.57 im/s)
[TRAIN] epoch 3/11,         observation 500/1899 batch,         loss: 0.0693 (avg 0.1375),         avg acc: 0.9593,         avg f1: 0.9597,         (21.34 im/s)
[TRAIN] epoch 3/11,      

 27%|██▋       | 3/11 [1:09:21<3:04:55, 1386.98s/it]


Epoch:  4
[TRAIN] epoch 4/11,         observation 0/1899 batch,         loss: 0.0158 (avg 0.0158),         avg acc: 1.0000,         avg f1: 1.0000,         (21.49 im/s)
[TRAIN] epoch 4/11,         observation 100/1899 batch,         loss: 0.1548 (avg 0.0686),         avg acc: 0.9827,         avg f1: 0.9842,         (21.59 im/s)
[TRAIN] epoch 4/11,         observation 200/1899 batch,         loss: 0.0139 (avg 0.0728),         avg acc: 0.9823,         avg f1: 0.9831,         (21.58 im/s)
[TRAIN] epoch 4/11,         observation 300/1899 batch,         loss: 0.5405 (avg 0.0813),         avg acc: 0.9782,         avg f1: 0.9789,         (21.43 im/s)
[TRAIN] epoch 4/11,         observation 400/1899 batch,         loss: 0.0232 (avg 0.0823),         avg acc: 0.9774,         avg f1: 0.9778,         (21.56 im/s)
[TRAIN] epoch 4/11,         observation 500/1899 batch,         loss: 0.0570 (avg 0.0844),         avg acc: 0.9768,         avg f1: 0.9771,         (21.61 im/s)
[TRAIN] epoch 4/11,      

 36%|███▋      | 4/11 [1:32:24<2:41:37, 1385.40s/it]

Epoch 4 validation acc: 0.9298, f1: 0.9276, loss: 0.2289 


Epoch:  5
[TRAIN] epoch 5/11,         observation 0/1899 batch,         loss: 0.0420 (avg 0.0420),         avg acc: 1.0000,         avg f1: 1.0000,         (21.48 im/s)
[TRAIN] epoch 5/11,         observation 100/1899 batch,         loss: 0.0247 (avg 0.0617),         avg acc: 0.9802,         avg f1: 0.9793,         (21.41 im/s)
[TRAIN] epoch 5/11,         observation 200/1899 batch,         loss: 0.0166 (avg 0.0584),         avg acc: 0.9820,         avg f1: 0.9814,         (21.44 im/s)
[TRAIN] epoch 5/11,         observation 300/1899 batch,         loss: 0.1730 (avg 0.0551),         avg acc: 0.9840,         avg f1: 0.9836,         (21.40 im/s)
[TRAIN] epoch 5/11,         observation 400/1899 batch,         loss: 0.0576 (avg 0.0580),         avg acc: 0.9832,         avg f1: 0.9831,         (21.67 im/s)
[TRAIN] epoch 5/11,         observation 500/1899 batch,         loss: 0.0178 (avg 0.0552),         avg acc: 0.9842,         avg

 45%|████▌     | 5/11 [1:55:27<2:18:26, 1384.50s/it]


Epoch:  6
[TRAIN] epoch 6/11,         observation 0/1899 batch,         loss: 0.0138 (avg 0.0138),         avg acc: 1.0000,         avg f1: 1.0000,         (21.27 im/s)
[TRAIN] epoch 6/11,         observation 100/1899 batch,         loss: 0.0101 (avg 0.0439),         avg acc: 0.9901,         avg f1: 0.9913,         (21.77 im/s)
[TRAIN] epoch 6/11,         observation 200/1899 batch,         loss: 0.0795 (avg 0.0408),         avg acc: 0.9897,         avg f1: 0.9902,         (21.58 im/s)
[TRAIN] epoch 6/11,         observation 300/1899 batch,         loss: 0.0070 (avg 0.0353),         avg acc: 0.9917,         avg f1: 0.9919,         (21.73 im/s)
[TRAIN] epoch 6/11,         observation 400/1899 batch,         loss: 0.1077 (avg 0.0339),         avg acc: 0.9921,         avg f1: 0.9923,         (21.89 im/s)
[TRAIN] epoch 6/11,         observation 500/1899 batch,         loss: 0.0187 (avg 0.0343),         avg acc: 0.9910,         avg f1: 0.9910,         (21.66 im/s)
[TRAIN] epoch 6/11,      

 55%|█████▍    | 6/11 [2:18:21<1:55:04, 1380.85s/it]

Epoch 6 validation acc: 0.9351, f1: 0.9336, loss: 0.2435 


Epoch:  7
[TRAIN] epoch 7/11,         observation 0/1899 batch,         loss: 0.0112 (avg 0.0112),         avg acc: 1.0000,         avg f1: 1.0000,         (21.71 im/s)
[TRAIN] epoch 7/11,         observation 100/1899 batch,         loss: 0.0520 (avg 0.0318),         avg acc: 0.9938,         avg f1: 0.9926,         (21.62 im/s)
[TRAIN] epoch 7/11,         observation 200/1899 batch,         loss: 0.0105 (avg 0.0283),         avg acc: 0.9944,         avg f1: 0.9940,         (21.52 im/s)
[TRAIN] epoch 7/11,         observation 300/1899 batch,         loss: 0.0060 (avg 0.0249),         avg acc: 0.9944,         avg f1: 0.9941,         (21.71 im/s)
[TRAIN] epoch 7/11,         observation 400/1899 batch,         loss: 0.0026 (avg 0.0276),         avg acc: 0.9938,         avg f1: 0.9933,         (21.80 im/s)
[TRAIN] epoch 7/11,         observation 500/1899 batch,         loss: 0.0327 (avg 0.0254),         avg acc: 0.9943,         avg

 64%|██████▎   | 7/11 [2:41:19<1:31:59, 1379.92s/it]


Epoch:  8
[TRAIN] epoch 8/11,         observation 0/1899 batch,         loss: 0.0149 (avg 0.0149),         avg acc: 1.0000,         avg f1: 1.0000,         (21.03 im/s)
[TRAIN] epoch 8/11,         observation 100/1899 batch,         loss: 0.0021 (avg 0.0209),         avg acc: 0.9932,         avg f1: 0.9934,         (21.55 im/s)
[TRAIN] epoch 8/11,         observation 200/1899 batch,         loss: 0.0036 (avg 0.0210),         avg acc: 0.9938,         avg f1: 0.9936,         (21.53 im/s)
[TRAIN] epoch 8/11,         observation 300/1899 batch,         loss: 0.0077 (avg 0.0218),         avg acc: 0.9942,         avg f1: 0.9941,         (21.60 im/s)
[TRAIN] epoch 8/11,         observation 400/1899 batch,         loss: 0.1180 (avg 0.0196),         avg acc: 0.9949,         avg f1: 0.9948,         (21.55 im/s)
[TRAIN] epoch 8/11,         observation 500/1899 batch,         loss: 0.0170 (avg 0.0190),         avg acc: 0.9950,         avg f1: 0.9948,         (21.44 im/s)
[TRAIN] epoch 8/11,      

 73%|███████▎  | 8/11 [3:04:14<1:08:55, 1378.52s/it]

Epoch 8 validation acc: 0.9334, f1: 0.9317, loss: 0.2813 


Epoch:  9
[TRAIN] epoch 9/11,         observation 0/1899 batch,         loss: 0.0048 (avg 0.0048),         avg acc: 1.0000,         avg f1: 1.0000,         (21.71 im/s)
[TRAIN] epoch 9/11,         observation 100/1899 batch,         loss: 0.0027 (avg 0.0093),         avg acc: 0.9975,         avg f1: 0.9972,         (21.59 im/s)
[TRAIN] epoch 9/11,         observation 200/1899 batch,         loss: 0.0034 (avg 0.0126),         avg acc: 0.9960,         avg f1: 0.9960,         (21.49 im/s)
[TRAIN] epoch 9/11,         observation 300/1899 batch,         loss: 0.0022 (avg 0.0137),         avg acc: 0.9963,         avg f1: 0.9961,         (21.55 im/s)
[TRAIN] epoch 9/11,         observation 400/1899 batch,         loss: 0.0020 (avg 0.0139),         avg acc: 0.9964,         avg f1: 0.9963,         (21.49 im/s)
[TRAIN] epoch 9/11,         observation 500/1899 batch,         loss: 0.0068 (avg 0.0156),         avg acc: 0.9960,         avg

 82%|████████▏ | 9/11 [3:27:16<45:58, 1379.40s/it]  


Epoch:  10
[TRAIN] epoch 10/11,         observation 0/1899 batch,         loss: 0.0039 (avg 0.0039),         avg acc: 1.0000,         avg f1: 1.0000,         (21.01 im/s)
[TRAIN] epoch 10/11,         observation 100/1899 batch,         loss: 0.0018 (avg 0.0125),         avg acc: 0.9988,         avg f1: 0.9985,         (21.46 im/s)
[TRAIN] epoch 10/11,         observation 200/1899 batch,         loss: 0.0045 (avg 0.0119),         avg acc: 0.9975,         avg f1: 0.9975,         (21.52 im/s)
[TRAIN] epoch 10/11,         observation 300/1899 batch,         loss: 0.0032 (avg 0.0129),         avg acc: 0.9971,         avg f1: 0.9972,         (21.59 im/s)
[TRAIN] epoch 10/11,         observation 400/1899 batch,         loss: 0.1917 (avg 0.0143),         avg acc: 0.9970,         avg f1: 0.9972,         (21.50 im/s)
[TRAIN] epoch 10/11,         observation 500/1899 batch,         loss: 0.0031 (avg 0.0134),         avg acc: 0.9969,         avg f1: 0.9970,         (21.87 im/s)
[TRAIN] epoch 10/1

 91%|█████████ | 10/11 [3:50:14<22:59, 1379.04s/it]

Epoch 10 validation acc: 0.9342, f1: 0.9333, loss: 0.2945 


Epoch:  11
[TRAIN] epoch 11/11,         observation 0/1899 batch,         loss: 0.0129 (avg 0.0129),         avg acc: 1.0000,         avg f1: 1.0000,         (21.77 im/s)


 91%|█████████ | 10/11 [3:50:19<23:01, 1381.94s/it]


KeyboardInterrupt: ignored

fold 0: lr=3e-5, accumulation step = 5, epochs = 5
- Epoch 1 validation: acc: 0.9215, f1: 0.9565
- Epoch 2 validation: acc: 0.9225, f1: 0.9576
- Epoch 3 validation: acc: 0.9268, f1: 0.9596 
- Epoch 4 validation: acc: 0.9325, f1: 0.9633 
- Epoch 5 validation: acc: 0.9345, f1: 0.9647 
- Epoch 6 validation: acc: 0.9336, f1: 0.9639 

fold 1: lr=3e-5, accumulation step = 5, epochs = 5
- Epoch 1 validation: acc: 0.9323, f1: 0.9627
- Epoch 2 validation: acc: 0.9317, f1: 0.9628 
- Epoch 3 validation: acc: 0.9331, f1: 0.9637 
- Epoch 4 validation: acc: 0.9302, f1: 0.9621 
- Epoch 5 validation: acc: 0.9332, f1: 0.9638 
- Epoch 6 validation: acc: 0.9341, f1: 0.9643 

fold 2: lr=3e-5, accumulation step = 5, epochs = 5
- Epoch 1 validation: acc: 0.9191, f1: 0.9553 
- Epoch 2 validation: acc: 0.9273, f1: 0.9599 
- Epoch 3 validation: acc: 0.9283, f1: 0.9608 
- Epoch 4 validation: acc: 0.9334, f1: 0.9639 
- Epoch 5 validation: acc: 0.9334, f1: 0.9639 
- Epoch 6 validation: acc: 0.9353, f1: 0.9650 

fold 3: lr=3e-5, accumulation step = 5, epochs = 5
- Epoch 1 validation: acc: 0.9156, f1: 0.9529 
- Epoch 2 validation: acc: 0.9308, f1: 0.9625 
- Epoch 3 validation: acc: 0.9294, f1: 0.9612
- Epoch 4 validation: acc: 0.9357, f1: 0.9652 
- Epoch 5 validation: acc: 0.9354, f1: 0.9650 
- Epoch 6 validation: acc: 0.9345, f1: 0.9646 

fold 4: lr=3e-5, accumulation step = 5, epochs = 5
- Epoch 1 validation: acc: 0.9213, f1: 0.9570
- Epoch 2 validation: acc: 0.9268, f1: 0.9601 
- Epoch 3 validation: acc: 0.9238, f1: 0.9583
- Epoch 4 validation: acc: 0.9348, f1: 0.9646 
- Epoch 5 validation: acc: 0.9313, f1: 0.9627 
- Epoch 6 validation: acc: 0.9336, f1: 0.9640 

In [ ]:
# load data test
X_test = _load_pkl('PhoBERT_pretrain/X_test.pkl')
y_test = _load_pkl('PhoBERT_pretrain/y_test.pkl')
print(X_test.shape)
print(y_test[:5])



(50373, 256)
[8 8 8 8 8]


In [ ]:
from fairseq.models.roberta import RobertaModel
import torch
import numpy as np
import os

phoBERT = RobertaModel.from_pretrained('PhoBERT_base_fairseq', checkpoint_file='model.pt')
phoBERT.register_classification_head('new_task', num_classes=10)

1042301B [00:01, 954193.30B/s]
456318B [00:00, 549127.74B/s]


In [ ]:
from tqdm import tqdm

DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
preds = []
logits = []
phoBERT.to(DEVICE)
phoBERT.load_state_dict(torch.load(os.path.join('model_ckpt_full', 'model.bin')))
test_dataset = torch.utils.data.TensorDataset(torch.tensor(X_test, dtype=torch.long))
test_loader = torch.utils.data.DataLoader(test_dataset, batch_size=16, shuffle=False)
phoBERT.eval()
for i, (x_batch,) in tqdm(enumerate(test_loader), total=len(test_loader)):
  logits = phoBERT.predict('new_task', x_batch)
  probs = torch.exp(logits)
  y_pred = np.argmax(probs.detach().cpu().numpy(), axis=1)
  preds = np.concatenate([preds, y_pred])

100%|██████████| 3149/3149 [12:27<00:00,  4.21it/s]


In [ ]:
from sklearn import metrics

f1 = metrics.f1_score(y_test, preds, average="micro")
acc = metrics.accuracy_score(y_test, preds)
print('f1 score: ', f1)
print('accuracy: ', acc)

f1 score:  0.9076886427252695
accuracy:  0.9076886427252695
